contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -269,6 +300,56 @@ contract ERC2771Forwarder is EIP712, Nonces {
         }
     }
 
+    /**
+     * @dev Returns whether the target trusts this forwarder.
+     *
+     * This function performs a static call to the target contract calling the
+     * {ERC2771Context-isTrustedForwarder} function.
+     */
+    function _isTrustedByTarget(address target) private view returns (bool) {
+        bytes4 selector = ERC2771Context.isTrustedForwarder.selector;
+
+        // Calldata is always the same, and we can guarantee its length is 36 bytes
+        // - 4 bytes from the selector
+        // - 32 bytes from the abi encoded target address (padded to 32)
+        uint8 calldataLength = 36;
+
+        bool success;
+        bool returnValue;
+        /// @solidity memory-safe-assembly
+        assembly {
+            // Because the return value is a boolean (requires 1 byte copied to memory) and the
+            // calldata length is 24 bytes, we can safely reuse the scratch space in memory (0x00 - 0x3F).
+            // This avoids memory leakage of allocating the encoded calldata in memory during each
+            // `isTrustedForwarder` call.
+            let ptr := 0x00
+
+            // | Location  | Content  | Content (Hex)                                                      |
+            // |-----------|----------|--------------------------------------------------------------------|
+            // |           |          |                                                  selector ↓        |
+            // | 0x00:0x1F | selector | 0x00000000000000000000000000000000000000000000000000000000AAAAAAAA |
+            // |           |          |   ↓ PADDING        target ↓                                        |
+            // | 0x20:0x3F | target   | 0x000000000000000000000000BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB |
+            mstore(ptr, shr(224, selector))
+            mstore(add(ptr, 0x20), address())
+
+            // Perform the staticcal and save the result in the scratch space.
+            // | Location  | Content  | Content (Hex)                                                      |
+            // |-----------|----------|--------------------------------------------------------------------|
+            // |           |          |                                                           result ↓ |
+            // | 0x00:0x1F | selector | 0x0000000000000000000000000000000000000000000000000000000000000001 |
+            success := staticcall(gas(), target, add(ptr, 0x1c), calldataLength, 0, 0x20)
+
+            // Avoid strange returndatasizes (eg. an EOA)
+            if eq(returndatasize(), 0x20) {
+                // Copy from memory and clean the `returnValue` as a boolean.
+                returnValue := shr(255, shl(255, mload(0)))
+            }
+        }
+
+        return success && returnValue;
+    }
```

Comment:
Do we really need assembly. Whenever possible, I tink we should stick with assembly. If solidity doiesn't provide the guarantees we need, then we should raise an issue.

Reply:
I guess the issue with 
```solidity
(bool success, bytes memory returndata) = target.staticcall(abi.encodeCall(ERC2771Context.isTrustedForwarder, (address(this))));
return success && returndata.length >= 0x20 && abi.decode(returndata, (bool));
```
is the returndata bomb

And the issue with 
```solidity
        if (target.code.length == 0) {
            return false;
        }
        try ERC2771Context(target).isTrustedForwarder(address(this)) returns (bool trusted) {
            return trusted;
        } catch {
            return false;
        }
```
if that target may return nothing ...

Reply:
https://github.com/ethereum/solidity/issues/14467

Reply:
Yeah, is a combination of those but also to avoid the memory leak we've seen in `Address` with the custom `string memory errorMsg` I think. @frangio mentioned it but I never found the issue (never knew if it was in Github)

Reply:
We do not need assembly for encoding the data though.

Reply:
@ernestognw @Amxx Please see https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4502/commits/889c11271f704e25b6d5685e9b0b8c36e4695fd3 and let me know what you think.

Reply:
IMO the less assembly the better !